datafusion/functions/src/macros.rs

```diff
@@ -226,9 +226,8 @@ macro_rules! make_math_unary_udf {
                     $EVALUATE_BOUNDS(inputs)
                 }
 
-                fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
-                    let args = ColumnarValue::values_to_arrays(args)?;
-
+                fn invoke(&self, col_args: &[ColumnarValue]) -> Result<ColumnarValue> {
```

Comment:
Note that @simonvandel and @jonahgao  have been removing use of `make_math_unary_udf` in general

(or example https://github.com/apache/datafusion/pull/12889)

We should probably make sure we aren't missing a similar problem in their reviews

Reply:
Is it necessary to add an attribute like `maintains_scalar` to `ScalarUDFImpl` and then automatically enforce this process during [evaluation](https://github.com/apache/datafusion/blob/d9450da6991a2977c617dfa789f208c53ba11421/datafusion/physical-expr/src/scalar_function.rs#L145)?

Reply:
I guess it's not necessarily true for volatile functions. But I wonder then how will the function know the array length? E.g. for a `rand(max)` function if I call it like `rand(42)` which is pretty reasonable, how does it work?

Reply:
If a function needs to maintain scalar, we can call `ColumnarValue::from_args_and_result` after invoking it. 
```rust
        // in file: datafusion/physical-expr/src/scalar_function.rs
        // evaluate the function
        let output = match self.args.is_empty() {
            true => self.fun.invoke_no_args(batch.num_rows()),
            false => self.fun.invoke(&inputs),
        }?;

        if let ColumnarValue::Array(array) = &output {
            if self.fun.need_maintain_scalar() && !inputs.is_empty() {
                output = ColumnarValue::from_args_and_result(inputs, array)
            }
        }
```

Reply:
I mean that if the function is volatile but it has only scalar arguments, how does it know how many rows to produce?

Reply:
The result after invoking is a `ColumnarValue`, and we can determine the number of rows .If a function might produce multiple rows for scalar args, it might not be appropriate to declare it as maintain_scalar 🤔.

Reply:
I think some other functions might also be handled like in this PR, so perhaps consider introducing a more general approach. Let's merge this PR first, and maybe we can consider this issue further later.

Reply:
> If a function might produce multiple rows for scalar args, it might not be appropriate to declare it as maintain_scalar

Yes, but how will you actually implement it?

Reply:
My thought is to add a new method, `fn maintains_scalar() -> bool`, to the `ScalarUDFImpl` trait, which by default returns false, with some math functions returning true. Or add such a flag in `Signature`. Check such a flag and apply similar processing as in this PR to all functions in one place.

Reply:
So what I'm trying to say is imagine I want to implement a volatile function `rand_int(max: Int64) -> Int64`. How can I implement it correctly so that given a `ScalarValue` for `max` it will actually produce an array of random integers with the expected length?

Reply:
Additionally, if I understand correctly, `rand_int` seems more like a table function, and supporting it in expression positions is not simple, similar to `unnset`, which I think goes beyond the scope of this PR.

Reply:
Well simply the issue is that I don't know how what is the batch size expected to return from this function. Usually it's inferred by any of the argument's length. But if all arguments are scalar values, then I can return a scalar value and all is well. This works for immutable functions but not for volatile ones. See e.g.

```rust
impl ScalarUDFImpl for RandomFunc {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn name(&self) -> &str {
        "random"
    }

    fn signature(&self) -> &Signature {
        &self.signature
    }

    fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType> {
        Ok(Float64)
    }

    fn invoke(&self, _args: &[ColumnarValue]) -> Result<ColumnarValue> {
        not_impl_err!("{} function does not accept arguments", self.name())
    }

    fn invoke_no_args(&self, num_rows: usize) -> Result<ColumnarValue> {
        let mut rng = thread_rng();
        let values = std::iter::repeat_with(|| rng.gen_range(0.0..1.0)).take(num_rows);
        let array = Float64Array::from_iter_values(values);
        Ok(ColumnarValue::Array(Arc::new(array)))
    }
}
```

What if I wanted to implement `RandomFunc` with an upper bound? I can't use `invoke_no_args` because I have one argument and I can't use `invoke` because it doesn't tell me how many rows I should produce.

Reply:
According to the definition of a scalar function, it produces a single row output for each row of input.
https://github.com/apache/datafusion/blob/4a0b7688074880767e1e032400c1e97b7ffcb76e/datafusion/expr/src/udf.rs#L37

I think this example should use `invoke`, and the number of rows produced is the same as the number of rows in args. The max arg of the i-th output row is taken from the i-th row of args[0].

Reply:
I understand your example: you want each invocation to use the same upper bound for multiple rows, and the result of each row is random. It‘s a special scalar function. I am not sure if there is a better way, maybe adding an extra placeholder argument to indicate the number of rows, for example: `select rand_int(10, a) from t`;

Reply:
Yeah anyway, I was just trying to show that this is a deficiency of the UDF model. Ideally, `invoke` should also get the number of expected rows and then there would also be no need for a special `invoke_no_args`.

Regarding adding a flag with default `false`, that doesn't sound very useful. Maybe we can just assume that if the result of any immutable function is an array of size one, we might as well convert it to a scalar value.

Reply:
> Ideally, `invoke` should also get the number of expected rows and then there would also be no need for a special `invoke_no_args`.

Make sense to me. This makes it easy to support `rand_int`. It seems that currently all functions should return a scalar value when all their arguments are scalar, for example, math functions that do not use `make_math_unary_udf`. 

> Maybe we can just assume that if the result of any immutable function is an array of size one, we might as well convert it to a scalar value.

This can be a general solution. The disadvantage is that if the arguments are not real scalar, it will cause unnecessary overhead, as the final result of queries will still need to be converted back into arrays.

Reply:
@joroKr21 @alamb Filed a new issue for other functions #12959

Reply:
> This can be a general solution. The disadvantage is that if the arguments are not real scalar, it will cause unnecessary overhead, as the final result of queries will still need to be converted back into arrays.

Yeah, but it sound very unlikely that we will have a batch of one row. In any case, we can do the same check that all arguments are scalar 👍

Reply:
> Yeah, but it sounds very unlikely that we will have a batch of one row.

it's unlikely, but you must not assume that.

a trivial example is a source table with one row.

> we can do the same check that all arguments are scalar 👍

👍